src/main/java/hudson/matrix/MatrixBuild.java

```diff
@@ -350,15 +350,26 @@ protected Result doRun(BuildListener listener) throws Exception {
             MatrixProject p = getProject();
             PrintStream logger = listener.getLogger();
 
-            // give axes a chance to rebuild themselves
-            activeConfigurations = p.rebuildConfigurations(this);
+            try {
+                // lock the project
+                p.buildLock();
```

Comment:
Your right. I think only the `rebuildConfigurations` causes the problem and has no impact on `listUpAggregators`. They should be moved outside the lock.

If the lock should only be used inside `MatrixProject` a new function must be created that returns the active configuration and the deep copy of the axes. Because such a function also would require the build context for `rebuildConfigurations` I am not sure if it is a good solution